File: src/main/org/nlogo/api/RemoteFile.java
package org.nlogo.api;

// we shouldn't need "File." lampsvn.epfl.ch/trac/scala/ticket/1409 - ST 3/6/11

public strictfp class RemoteFile extends File {

  private final String filepath;
  private String suffix = null;
  private File.Mode mode = File.Mode.NONE;
  private java.io.BufferedReader buffReader;

  public RemoteFile(String filepath) {
    this.filepath = filepath;
  }

  public RemoteFile(String filepath, String suffix) {
    if (filepath != null && suffix != null) {
      String tmpf = filepath.toLowerCase();
      String tmps = suffix.toLowerCase();
      if (tmpf.endsWith(tmps)) {
        this.filepath = filepath.substring(0, tmpf.lastIndexOf(tmps));
      } else {
        this.filepath = filepath;
      }
    } else {
      this.filepath = filepath;
    }
    this.suffix = suffix;
  }

  @Override
  public File.Mode getMode() {
    return mode;
  }


  @Override
  public java.io.PrintWriter getPrintWriter() {
    return null;
  }

  @Override
  public java.io.BufferedReader getBufferedReader() {
    return buffReader;
  }

  @Override
  public java.io.InputStream getInputStream()
      throws java.io.IOException {
    return new java.io.BufferedInputStream
        (new java.net.URL
            (org.nlogo.util.Utils.escapeSpacesInURL(getPath()))
            .openStream());
  }

  public static boolean exists(String path) {
    String url = org.nlogo.util.Utils.escapeSpacesInURL(path);
    try {
      new java.net.URL(url).openStream();
      return true;
    } catch (java.io.IOException ex) {
      return false;
    }

  }

  @Override
  public void open(File.Mode mode)
      throws java.io.IOException {
    if (buffReader != null) {
      throw new java.io.IOException("Attempted to open an already open file");
    }
    // Comment out this line and switch lines below to enable renaming code
    String fullpath = suffix == null ? filepath : filepath + suffix;
    switch (mode) {
      case READ:
        pos = 0;
        eof = false;
        buffReader = new java.io.BufferedReader(
            new java.io.InputStreamReader(
                new java.io.BufferedInputStream(
                    new java.net.URL(org.nlogo.util.Utils.escapeSpacesInURL(fullpath)).openStream())));
        this.mode = mode;
        break;
      case WRITE:
        throw new java.io.IOException("Cannot write to remote files.");
      case APPEND:
        throw new java.io.IOException("Cannot append to remote files.");
      default:
        break;
    }
  }

  @Override
  public void print(String str)
      throws java.io.IOException {
    throw new java.io.IOException("Attempted to print to an unopened File");
  }

  @Override
  public void println(String line)
      throws java.io.IOException {
    throw new java.io.IOException("Attempted to println to an unopened File");
  }

  @Override
  public void println()
      throws java.io.IOException {
    throw new java.io.IOException("Attempted to println to an unopened File");
  }

  @Override
  public void flush() {
  }

  @Override
  public void close(boolean ok)
      throws java.io.IOException {
    if (buffReader == null) {
      return;  // not an error
    }
    switch (mode) {
      case WRITE:
      case APPEND:
        break;
      case READ:
        buffReader.close();
        buffReader = null;
        break;
      default:
        break;
    }
    mode = File.Mode.NONE;

  }

  @Override
  public String getAbsolutePath() {
    return filepath;
  }

  @Override
  public String getPath() {
    return filepath;
  }

}

File: src/main/org/nlogo/api/RemoteFile.scala
package org.nlogo.api

object RemoteFile {
  def exists(path: String): Boolean = {
    val url = org.nlogo.util.Utils.escapeSpacesInURL(path)
    try { new java.net.URL(url).openStream(); true }
    catch {
      case ex: java.io.IOException =>
        false
    }
  }
}

class RemoteFile(_filepath: String, suffix: String) extends File {

  def this(_filepath: String) = this(_filepath, null)

  val filepath =
    if (_filepath == null || suffix == null)
      _filepath
    else {
      val tmpf = _filepath.toLowerCase
      val tmps = suffix.toLowerCase
      if (tmpf.endsWith(tmps))
        _filepath.substring(0, tmpf.lastIndexOf(tmps))
      else
        _filepath
    }

  override def getPrintWriter = null

  @throws(classOf[java.io.IOException])
  override def getInputStream =
    new java.io.BufferedInputStream(
      new java.net.URL(
        org.nlogo.util.Utils.escapeSpacesInURL(getPath))
      .openStream())

  @throws(classOf[java.io.IOException])
  override def open(mode: FileMode) {
    if (reader != null)
      throw new java.io.IOException(
        "Attempted to open an already open file")
    val fullpath =
      filepath + Option(suffix).getOrElse("")
    mode match {
      case FileMode.READ =>
        pos = 0
        eof = false
        reader = new java.io.BufferedReader(
            new java.io.InputStreamReader(
                new java.io.BufferedInputStream(
                    new java.net.URL(org.nlogo.util.Utils.escapeSpacesInURL(fullpath)).openStream())))
        this.mode = mode
      case FileMode.WRITE | FileMode.APPEND =>
        unsupported
    }
  }

  private def unsupported = 
    throw new java.io.IOException("Cannot write to remote files.")

  @throws(classOf[java.io.IOException])
  override def print(str: String) = unsupported
  @throws(classOf[java.io.IOException])
  override def println(line: String) = unsupported
  @throws(classOf[java.io.IOException])
  override def println() = unsupported

  override def flush() { }

  @throws(classOf[java.io.IOException])
  override def close(ok: Boolean) {
    mode match {
      case FileMode.READ =>
        reader.close()
        reader = null
      case FileMode.NONE =>
    }
    mode = FileMode.NONE
  }

  override def getAbsolutePath =
    new java.io.File(filepath + Option(suffix).getOrElse("")).getAbsolutePath

  override def getPath =
    new java.io.File(filepath + Option(suffix).getOrElse("")).getPath

}
